lib/using_restricted_array.rb

```diff
@@ -1,61 +1,126 @@
+# Time complexity: would be O(n) we have to iterate through every single index in order to count it and
+# We don't know how long the data structure is going to be.
+# Space complexity: O(1) There are no added data structures being added besides the counter variable.
 def length(array)
-  raise NotImplementedError
+  if array != []
+    counter = 0
+    i = 0
+    until array[i] == nil
+      i += 1
+      counter += 1
+    end
+    return counter
+  end
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: This would be O(n) becuase it iterates through an array printing out each value and we don't know how long the array will be.
+# Space complexity: This would be O(1) becuase it there are no continual additions of data structures
 def print_array(array)
-  raise NotImplementedError
+  length = length(array)
+  string = ""
+  i = 0
+  while i < length
+    string += array[i] + " "
+    i += 1
+  end
+  print(string)
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(logn)
+# Space complexity: O(1)
+
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  i = 0
+  while i < length
+    if array[i] == value_to_find
+      return true
+    else
+      i += 1
+    end
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
+
 def find_largest(array, length)
-  raise NotImplementedError
+  largest = array[0]
+  i = 0
+  while i < length
+    if array[i] > largest
+      largest = array[i]
+      i += 1
+    end
+    i += 1
+  end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest = array[0]
+  i = 0
+  while i < length
+    if array[i] < smallest
+      smallest = array[i]
+      i += 1
+    end
+    i += 1
+  end
+  return smallest
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  iterations = length / 2
+  iterations.times do |index|
+    temporary = array[index]
+    array[index] = array[length - 1 - index]
+    array[length - 1 - index] = temporary
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(logn)
+# Space complexity: O(1)
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  bigger = array.length - 1
```

Comment:
In restricted Array, length is not a method.  It's a parameter passed in.

```suggestion
  bigger = length - 1
```